scipy/sparse/_coo.py

```diff
@@ -155,7 +155,7 @@ def __init__(self, arg1, shape=None, dtype=None, copy=False):
                     M, N = shape
                     self._shape = check_shape((M, N))
 
-                idx_dtype = get_index_dtype(maxval=max(self.shape))
+                idx_dtype = self._get_index_dtype((row, col), maxval=max(self.shape), check_contents=True)
```

Comment:
We weren't doing check_contents before. Why add that here?

Reply:
We're doing this here because `(row, col)` are now being passed. In order to let the dtype of these propagate, but also downcast for the matrix case I needed to check the values.

The change of adding `(row, col)` before I added `check_contents` was causing many test failures due to `bsr.tocoo()` returning some `int64` indices, related to: https://github.com/scipy/scipy/pull/18509#discussion_r1203071761